### PR TITLE
Speeds up blur script so it's usable

### LINKED ISSUE #5836: Blurring causes freezing
Blurring beta functionality doesn't work with Google Chrome causing freezing the tab totally.

### PR DESCRIPTION
**Pull request by jbis9051**
Fixes #5836

### CODE REVIEW

**Comment by jbis9051** [comment]
Ok. I tried to fix all the errors. I don't know why 

```ts
 dataL[(n * 4) + 1] = this.imageData.data[(n * 4) + 1];
    dataL[(n * 4) + 2] = this.imageData.data[(n * 4) + 2];
    dataL[(n * 4) + 3] = this.imageData.data[(n * 4) + 3];
}
```

is an issue when a couple lines before, I make sure `this.imageData` is not null.

```
if (this.imageData) {
```

**Comment by jbis9051** [comment]
Alright i fixed it, sort of

**Comment by jbis9051** [comment]
Signed the doc

**Comment by daxiondi** [comment]
@saghul please review.

**Comment by heikar** [comment]
Any news?
I am of the same opinion as @daxiondi. I think this is an important part of jitsi, so @saghul and the Team, please review and publish it.

**Comment by audax** [comment]
You need to return the stream from your `startEffect()` method and you need a `isEnabled()` method. Otherwise `node_modules/lib-jitsi-meet/modules/RTC/JitsiLocalTrack.js:386` is calling an undefined method and later uses an undefined video track.
```
Index: react/features/stream-effects/blur/JitsiStreamBlurEffect.js
<+>UTF-8
===================================================================
--- react/features/stream-effects/blur/JitsiStreamBlurEffect.js	(revision de5539a4162dbe8ba662f84e9dc8fc311def8b06)
+++ react/features/stream-effects/blur/JitsiStreamBlurEffect.js	(date 1600076911524)
@@ -92,6 +92,7 @@
 
         // Workaround for FF issue https://bugzilla.mozilla.org/show_bug.cgi?id=1388974
         this.finalCanvas.getContext('2d');
+        return this.getStream()
     }
 
 
@@ -104,6 +105,18 @@
         this.shouldContinue = false;
     }
 
+
+
+    /**
+     * Checks if the JitsiLocalTrack supports this effect.
+     *
+     * @param {JitsiLocalTrack} sourceLocalTrack - Track to which the effect will be applied.
+     * @returns {boolean} - Returns true if this effect can run on the specified track, false otherwise.
+     */
+    isEnabled(jitsiLocalTrack: Object) {
+        return jitsiLocalTrack.isVideoTrack();
+    }
+
     /**
      * Render next frame.
      *
```

**Comment by jbis9051** [comment]
@audax I'll add a fix for the issue you mentioned. 

See here, for the image thing. https://github.com/jitsi/jitsi-meet/issues/5836#issuecomment-661632382

It's there for future use.

**Comment by jbis9051** [comment]
@audax see https://github.com/jitsi/jitsi-meet/pull/7473/commits/d2a1595e17e7f7ea154e4e5ce32b7991556878da

**Comment by audax** [comment]
This PR, maybe combined with the [green-screen PR](https://github.com/jitsi/jitsi-meet/pull/7345) could be a game changer for jitsi.

So, I can confirm that this PR does what it says on the box: The blur effect is less taxing on the CPU. I would remove the background-image feature however until there is a UI for it, so that the changes to the source are minimal.

**Comment by jbis9051** [comment]
@audax I agree. 

I still think background image feature should be included. There's no harm in including it and it makes it easier for anyone that wants to implement that feature.

P.S. This code came from my other project, BitLink. As such, I've implemented a UI for the image selector in the other app. You can see the component code [here](https://github.com/oss-videochat/bitlink/blob/master/frontend/src/components/Modals/Settings/SettingsViews/VideoEffects.tsx). If you want to test out the background replacement as a proof of concept of what jitsi could do for the ui go to https://bitlink.live/ > Create a Room > Gears in upper right > Camera Settings > New + and select an image. Background replacement is more performant than blur.

**Comment by rickgoud** [comment]
Just tested it on your Bitlink, and works Awesome!!
Would be great if the team could bring that to master, including virtual background selector. Awesome work here!!

**Comment by jallamsetty1** [comment]
@jbis9051, thank you so much for all this work. I do see that the blur effect is a little less taxing on the cpu now. However, the frame rate that we get when blur is enabled seems to remain the same as before, i.e., around 4-5 fps for a 720p stream. This feature is still in beta because we haven't been able to find time to explore other mechanisms to bring this frame rate closer to 30 fps so that video looks better. 
Having said that, we would like to get these changes in. Can you please separate out the background-image feature into a different PR so the changes are minimal here.
Here is a link to our coding guidelines - https://jitsi.github.io/handbook/docs/dev-guide/dev-guide-contributing. Can you please go over these and make some changes before we are able to review, thanks!

**Comment by jbis9051** [comment]
@jallamsetty1  

> Can you please separate out the background-image feature into a different PR so the changes are minimal here

I'll make those changes.

> However, the frame rate that we get when blur is enabled seems to remain the same as before, i.e., around 4-5 fps for a 720p stream

I can't check the exact framerate right now but I am getting a much higher frame rate than 4-5fps (on a 640 by 480).

Could you check on https://app.bitlink.live/create > Gears in upper right > Camera Settings > `blur`? I've not tested on jitsi itself, but that site is running almost an identical script. I'd like to see if its an issue with jitsi specifically or something else.

P.S. Just to clarify, I have not increased tensorflow's performance significantly. Instead, I changed the script from segmenting each frame to segmenting when the previous segmentation is done and in the mean time using the old segmentation data. This allows us to achieve a significantly higher frame rate out of the video itself without putting as much pressure on the cpu/gpu. However, tensorflow's fps is probably around the same. Are you testing tensorflow's fps itself or the framerate of the actual video?

**Comment by jbis9051** [comment]
@jallamsetty1 for reference here is what i am experiencing. This is a comparison of without blur and with blur enabled. Even with imgur degradation of frame rate due to compression it still looks much higher than 4-5fps . Looks around 18-20fps. Not perfect but much better than 4-5fps.

https://i.imgur.com/l1I3Wp7.mp4

**Comment by jallamsetty1** [comment]
> @jallamsetty1
> 
> > Can you please separate out the background-image feature into a different PR so the changes are minimal here
> 
> I'll make those changes.

Thank you !
> 
> > However, the frame rate that we get when blur is enabled seems to remain the same as before, i.e., around 4-5 fps for a 720p stream
> 
> I can't check the exact framerate right now but I am getting a much higher frame rate than 4-5fps (on a 640 by 480).
> 
> Could you check on `https://app.bitlink.live/create > Gears in upper right > Camera Settings > blur`? I've not tested on jitsi itself, but that site is running almost an identical script (see comment above). I'd like to see if its an issue with jitsi specifically or something else.

I checked it on your demo and I am still getting the same 4-5 fps with my Logitech Brio camera. 

> P.S. Just to clarify, I have not increased tensorflow's performance significantly. Instead, I changed the script from segmenting each frame to segmenting when the previous segmentation is done and in the mean time using the old segmentation data. This allows us to achieve a significantly higher frame rate out of the video itself without putting as much pressure on the cpu/gpu. However, tensorflow's fps is probably around the same. Are you testing tensorflow's fps itself or the framerate of the final video?

I am looking at the fps of the final video. We aren't segmenting each frame, we call the segmentation API when the previous segmentation is complete. I think using stack-blur instead of using BodyPix's Bokeh effect is what making the difference in the cpu consumption. The only drawback I see with stack-blur is that we will lose the ability to control the amount of blur that Bokeh effect provides. Will you be able to modify just the mask rendering part instead of re-writing everything ? Like we could re-use most of the variables, the renderMask method name makes more sense than "tick".

**Comment by jbis9051** [comment]
> I checked it on your demo and I am still getting the same 4-5 fps with my Logitech Brio camera.

This is quite odd. I've tested on multiple computers with multiple people. I just tested on latest Chrome on a 2013 PC with Windows 8.1, a Core i3, and 6GB of RAM. Despite the less than ideal specs, I got near real time speed.

> We aren't segmenting each frame, we call the segmentation API when the previous segmentation is complete. I think using stack-blur instead of using BodyPix's Bokeh effect is what making the difference in the cpu consumption.  

I apologize, I didn't explain it correctly. The key part is "and  in the mean time using the old segmentation data". In Jitsi's current code if the previous segmentation isn't complete (`!this._maskInProgress`), the `_onMaskFrameTimer` doesn't call `_renderMask` and wait's for the worker to trigger another check. This means that Jitsi only actually renders a frame as fast as tensorflow can process them. See https://github.com/jitsi/jitsi-meet/blob/master/react/features/stream-effects/blur/JitsiStreamBlurEffect.js#L57.

My script does the same check, but if the previous segmentation isn't complete it **still** renders out using the old segmentation data which is usually *close enough* unless tensorflow is very slow and the user is move around a bunch. See [here](https://github.com/jitsi/jitsi-meet/pull/7473/files#diff-f86527521031a65bbcc36708deb32b22R126) for the check. Regardless of the outcome, [L133](https://github.com/jitsi/jitsi-meet/pull/7473/files#diff-f86527521031a65bbcc36708deb32b22R133) is called and a frame is rendered.

The stack-blur is definitively the reason the cpu usage is less but the frame rate should still be much higher.

> The only drawback I see with stack-blur is that we will lose the ability to control the amount of blur that Bokeh effect provides.

You can modify the amount of blur. I have it set to 12 but you can modify. See [here](https://github.com/jitsi/jitsi-meet/pull/7473/files#diff-f86527521031a65bbcc36708deb32b22R157) and [the docs](https://github.com/flozz/StackBlur#api)

>  Will you be able to modify just the mask rendering part instead of re-writing everything ? Like we could re-use most of the variables, the renderMask method name makes more sense than "tick".

I rewrote everything because I pulled from a different project which also had background replacement. (Sidenote, I also tested with a worker but found it did not solve the issue of browsers throttling when window goes out of focus so I didn't include it).


If you would much prefer, I can revert back to Jitis's script and modify it minimally to use stackblur and trick aforementioned, however it will be much easier for me to create another PR with the image replacement if we stick with this script. I can also rename the functions to be more consistent with Jtisi's previous code.

**Comment by jallamsetty1** [comment]
> @jallamsetty1
> 
> > I checked it on your demo and I am still getting the same 4-5 fps with my Logitech Brio camera.
> 
> This is quite odd. I've tested on multiple computers with multiple people. I just tested on latest Chrome on a 2013 PC with Windows 8.1, a Core i3, and 6GB of RAM. Despite the less than ideal specs, I got near real time speed.
> 
> Also see [#7473 (comment)](https://github.com/jitsi/jitsi-meet/pull/7473#issuecomment-692158256) and [#7473 (comment)](https://github.com/jitsi/jitsi-meet/pull/7473#issuecomment-692136600) neither mention an fps issue
> 
> > We aren't segmenting each frame, we call the segmentation API when the previous segmentation is complete. I think using stack-blur instead of using BodyPix's Bokeh effect is what making the difference in the cpu consumption.
> 
> I apologize, I didn't explain it correctly. The key part is "and in the mean time using the old segmentation data". In Jitsi's current code if the previous segmentation isn't complete (`!this._maskInProgress`), the `_onMaskFrameTimer` doesn't call `_renderMask` and wait's for the worker to trigger another check. This means that Jitsi only actually renders a frame as fast as tensorflow can process them. See https://github.com/jitsi/jitsi-meet/blob/master/react/features/stream-effects/blur/JitsiStreamBlurEffect.js#L57.
> 
> My script does the same check, but if the previous segmentation isn't complete it **still** renders out using the old segmentation data which is usually _close enough_ unless tensorflow is very slow and the user is move around a bunch. See [here](https://github.com/jitsi/jitsi-meet/pull/7473/files#diff-f86527521031a65bbcc36708deb32b22R126) for the check. Regardless of the outcome, [L133](https://github.com/jitsi/jitsi-meet/pull/7473/files#diff-f86527521031a65bbcc36708deb32b22R133) is called and a frame is rendered.

Yes, I understand that part and that was deliberate since performance was so bad in an earlier version that using old data didn't produce good results whenever there was motion involved. 
> 
> The stack-blur is definitively the reason the cpu usage is less but the frame rate should still be much higher.
> 
> > The only drawback I see with stack-blur is that we will lose the ability to control the amount of blur that Bokeh effect provides.
> 
> You can modify the amount of blur. I have it set to 12 but you can modify. See [here](https://github.com/jitsi/jitsi-meet/pull/7473/files#diff-f86527521031a65bbcc36708deb32b22R157) and [the docs](https://github.com/flozz/StackBlur#api)
That is great !
> 
> > Will you be able to modify just the mask rendering part instead of re-writing everything ? Like we could re-use most of the variables, the renderMask method name makes more sense than "tick".
> 
> I rewrote everything because I pulled from a different project which also had background replacement. (Sidenote, I also tested with a worker but found it did not solve the issue of browsers throttling when window goes out of focus so I didn't include it).
> 
> If you would much prefer, I can revert back to Jitis's script and modify it minimally to use stackblur and trick aforementioned, however it will be much easier for me to create another PR with the image replacement if we stick with this script. I can also rename the functions to be more consistent with Jtisi's previous code.

This would be great, let's keep changes to a minimum - switch to using stackblur and the re-use old data trick. 
Have you explored moving everything to a WebWorker (not just the timer) ? Apparently this is a trick that many use, please see https://github.com/tensorflow/tfjs/issues/2723#issuecomment-655816617

**Comment by jbis9051** [comment]
> This would be great, let's keep changes to a minimum - switch to using stackblur and the re-use old data trick.

Will do soon.

**Comment by jbis9051** [comment]
> Have you explored moving everything to a WebWorker (not just the timer) ? Apparently this is a trick that many use, please see tensorflow/tfjs#2723 (comment)

I did minimal testing with Chrome (as that seems to be the only browser that supports the OffscreenCanvas). I didn't see much of a difference, but again it was very minimal testing (one computer, once, eyeballing it) so we could try it out in the future. I suspect within a couple years all browser will support it.

**Comment by jallamsetty1** [comment]
> > Have you explored moving everything to a WebWorker (not just the timer) ? Apparently this is a trick that many use, please see [tensorflow/tfjs#2723](https://github.com/tensorflow/tfjs/issues/2723) (comment)
> 
> I did minimal testing with Chrome (as that seems to be the only browser that supports the OffscreenCanvas). I didn't see much of a difference, but again it was very minimal testing (one computer, once, eyeballing it) so we could try it out in the future. I suspect within a couple years all browser will support it.

Thanks for trying it out. I am surprised it isn't helping. Maybe we would see a difference when the machine is loaded. There is a performance metric we have called LongTask events which is a good indicator of tasks getting backed up in queue if the system is loaded. I could probably do some more digging later to see if we there is a difference in that metric w/ and w/o WebWorker.

**Comment by jbis9051** [comment]
@jallamsetty1 I've reverted back to the old script and made minimal changes. I also changed some things around with the worker. Seems to work (no pun intended) better.

I am still getting better speed on BitLink for some reason. Not exactly sure why. It could just be that Jitsi is a much heavier application and therefore runs slower.

File: react/features/stream-effects/blur/JitsiStreamBlurEffect.js [hunk 1]
```diff
@@ -43,6 +44,10 @@ export default class JitsiStreamBlurEffect {
         this._outputCanvasElement = document.createElement('canvas');
         this._outputCanvasElement.getContext('2d');
         this._inputVideoElement = document.createElement('video');
+        this._inputVideoCanvasElement = document.createElement('canvas');
+
+        this._maskFrameTimerWorker = new Worker(timerWorkerScript, { name: 'Blur effect worker' });
+        this._maskFrameTimerWorker.onmessage = this._onMaskFrameTimer;
```
Comment (jallamsetty1) [inline react/features/stream-effects/blur/JitsiStreamBlurEffect.js:50]: Is there a reason why this was moved up to the constructor from the `startEffect` ? The idea was to create the Worker only when the effect is enabled and terminate the Worker in `stopEffect`. Does the blur work fine when its disabled and enabled a second time ?

**Comment by jallamsetty1** [comment]
> @jallamsetty1 I've reverted back to the old script and made minimal changes. I also changed some things around with the worker. Seems to work (no pun intended) better.
> 
> I am still getting better speed on BitLink for some reason. Not exactly sure why. It could just be that Jitsi is a much heavier application and therefore runs slower.

Thanks @jbis9051, the changes look good to me! I had a question though, please check my comment above.

File: react/features/stream-effects/blur/JitsiStreamBlurEffect.js (hunk 1 above)
Comment (jbis9051) [reply react/features/stream-effects/blur/JitsiStreamBlurEffect.js]: Mistake left over from testing. Removed.

**Comment by jbis9051** [comment]
@jallamsetty1 I've corrected the issue. Thank you!

**Review by jallamsetty1 — APPROVED** [review]

**Comment by jallamsetty1** [comment]
> @jallamsetty1 I've corrected the issue. Thank you!
> 
> I've also changed the title. I will save the virtual background/background replacement for a new PR.

Can you pleas rebase the branch to fix the conflicts ? Also pls sign the CLA if you haven't done so already. thanks !

File: package.json [hunk 2]
```diff
@@ -91,6 +91,7 @@
     "redux-thunk": "2.2.0",
     "rnnoise-wasm": "github:jitsi/rnnoise-wasm.git#566a16885897704d6e6d67a1d5ac5d39781db2af",
     "rtcstats": "github:jitsi/rtcstats#v6.2.0",
+    "stackblur-canvas": "^2.3.0",
```
Comment (saghul) [inline package.json:94]: Please drop the caret (^)

Comment (jbis9051) [reply package.json]: https://github.com/jitsi/jitsi-meet/pull/7473/commits/cc3231813be0920417f28aef1895d53177ed4616

**Comment by jbis9051** [comment]
@jallamsetty1 

> Can you pleas rebase the branch to fix the conflicts ?

Shouldn't be any conflicts now. Let me know if i screwed it up.

**Review by jallamsetty1 — APPROVED** [review]

**Comment by jallamsetty1** [comment]
> @jallamsetty1
> 
> > Can you pleas rebase the branch to fix the conflicts ?
> 
> Shouldn't be any conflicts now. Let me know if i screwed it up.

Looks like we still have the conflicts.

**Comment by saghul** [comment]
@jallamsetty1 you can squash-merge.

**Comment by jallamsetty1** [comment]
Jenkins test this please.

**Comment by jbis9051** [comment]
@saghul For future reference, was the merge necessary?  https://github.com/jitsi/jitsi-meet/pull/7473/commits/3e3d0295ae0efc7198bd15c43e46c4d380c03a15

**Comment by saghul** [comment]
@jbis9051 Best would have been a rebase, but we can squash-merge so that's fine. Thanks for your contribution!
